File: Integrations/LiveAction/LiveNX-Inventory-To-ThirdEye/Python/live_nx_to_thirdeye.py
#!/usr/bin/env python3

import argparse
import copy
import csv
import io
import ipaddress
import json
import os
import sys
import uuid
from dataclasses import dataclass

import requests
from dotenv import load_dotenv


class BridgeError(RuntimeError):
    pass


def env_bool(name, default=False):
    value = os.environ.get(name)
    if value is None:
        return default
    return value.strip().lower() in {"1", "true", "yes", "on"}


def require_env(name):
    value = os.environ.get(name)
    if not value:
        raise BridgeError(f"Set {name} in .env before running this integration.")
    return value


def normalized_header(value):
    return "".join(character for character in value.upper() if character.isalnum())


def sorted_ip_addresses(addresses):
    return sorted(
        addresses,
        key=lambda value: (
            ipaddress.ip_address(value).version,
            ipaddress.ip_address(value),
        ),
    )


def parse_live_nx_device_ips(csv_text, require_vendor=True):
    reader = csv.DictReader(io.StringIO(csv_text.lstrip("\ufeff")))
    if not reader.fieldnames:
        raise BridgeError("The LiveNX device export did not contain a CSV header.")

    fields = {normalized_header(name): name for name in reader.fieldnames if name}
    ip_field = next(
        (
            fields[name]
            for name in ("IPADDRESS", "MANAGEMENTIPADDRESS", "MANAGEMENTIP", "IP")
            if name in fields
        ),
        None,
    )
    if ip_field is None:
        available = ", ".join(reader.fieldnames)
        raise BridgeError(
            "The LiveNX CSV does not contain a recognized IP-address column. "
            f"Available columns: {available}"
        )

    vendor_field = fields.get("VENDOR")
    if require_vendor and vendor_field is None:
        raise BridgeError(
            "LIVENX_REQUIRE_VENDOR=true, but the LiveNX CSV has no VENDOR column."
        )

    addresses = set()
    for row in reader:
        if require_vendor and not (row.get(vendor_field) or "").strip():
            continue

        candidate = (row.get(ip_field) or "").strip()
        try:
            addresses.add(str(ipaddress.ip_address(candidate)))
        except ValueError:
            continue

    return addresses


def prepare_discovery_job(job_data, network, addresses):
    prepared = copy.deepcopy(job_data)
    parameters = prepared.get("jobParameters")
    if not isinstance(parameters, dict):
        raise BridgeError("The selected job has no jobParameters object.")
    if "includedAddresses" not in parameters:
        raise BridgeError(
            "The selected job is not a compatible Discover Devices job: "
            "jobParameters.includedAddresses is missing."
        )

    if "managedNetwork" in prepared:
        prepared["managedNetwork"] = network
    if "managedNetworks" in prepared:
        current = prepared["managedNetworks"]
        prepared["managedNetworks"] = [network] if isinstance(current, list) else network
    if "managedNetwork" in parameters:
        parameters["managedNetwork"] = network

    parameters["includedAddresses"] = ",".join(sorted_ip_addresses(addresses))
    return prepared


@dataclass(frozen=True)
class BridgeConfig:
    live_nx_base_url: str
    live_nx_api_token: str
    live_nx_export_path: str
    netld_base_url: str
    netld_api_key: str
    netld_network: str
    discovery_job_name: str | None
    timeout: float
    require_vendor: bool
    debug: bool

    @classmethod
    def from_env(cls):
        load_dotenv()
        try:
            timeout = float(os.environ.get("REQUEST_TIMEOUT_SECONDS", "30"))
        except ValueError as error:
            raise BridgeError("REQUEST_TIMEOUT_SECONDS must be a number.") from error
        if timeout <= 0:
            raise BridgeError("REQUEST_TIMEOUT_SECONDS must be greater than zero.")

        return cls(
            live_nx_base_url=require_env("LIVENX_BASE_URL").rstrip("/"),
            live_nx_api_token=require_env("LIVENX_API_TOKEN"),
            live_nx_export_path=os.environ.get(
                "LIVENX_DEVICE_EXPORT_PATH", "/v1/devices/export/csv"
            ),
            netld_base_url=require_env("NETLD_BASE_URL").rstrip("/"),
            netld_api_key=require_env("NETLD_API_KEY"),
            netld_network=os.environ.get("NETLD_NETWORK", "Default"),
            discovery_job_name=os.environ.get("NETLD_DISCOVERY_JOB_NAME"),
            timeout=timeout,
            require_vendor=env_bool("LIVENX_REQUIRE_VENDOR", True),
            debug=env_bool("NETLD_DEBUG", False),
        )


class LiveNXClient:
    def __init__(self, base_url, api_token, export_path, timeout: float = 30):
        self.base_url = base_url.rstrip("/")
        self.api_token = api_token
        self.export_path = "/" + export_path.lstrip("/")
        self.timeout = timeout
        self.session = requests.Session()

    def export_devices_csv(self):
        url = f"{self.base_url}{self.export_path}"
        try:
            response = self.session.get(
                url,
                headers={
                    "Accept": "text/csv",
                    "Authorization": f"Bearer {self.api_token}",
                },
                timeout=self.timeout,
                verify=True,
                allow_redirects=False,
            )
        except requests.RequestException as error:
            raise BridgeError(
                f"Could not retrieve the LiveNX device export from {url}."
            ) from error

        if response.is_redirect:
            raise BridgeError(
                "The LiveNX export request was redirected. Confirm LIVENX_BASE_URL, "
                "LIVENX_DEVICE_EXPORT_PATH, and the API token."
            )
        try:
            response.raise_for_status()
        except requests.HTTPError as error:
            raise BridgeError(
                f"LiveNX device export failed with HTTP {response.status_code}."
            ) from error

        response.encoding = response.encoding or "utf-8"
        return response.text


class NetLDClient:
    def __init__(self, base_url, api_key, timeout: float = 30, debug=False):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.debug = debug
        self.session = requests.Session()

    @property
    def headers(self):
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    def login(self):
        try:
            response = self.session.get(
                f"{self.base_url}/rest",
                headers={"Authorization": f"Bearer {self.api_key}"},
                timeout=self.timeout,
                verify=True,
                allow_redirects=False,
            )
        except requests.RequestException as error:
            raise BridgeError(f"Could not reach {self.base_url}.") from error

        if response.is_redirect:
            raise BridgeError("ThirdEye login redirected instead of creating an API session.")
        try:
            response.raise_for_status()
        except requests.HTTPError as error:
            raise BridgeError(
                f"ThirdEye login failed with HTTP {response.status_code}."
            ) from error

    def call(self, method, **params):
        payload = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": str(uuid.uuid4()),
        }
        if self.debug:
            print(f"ThirdEye request: {json.dumps(payload, indent=2)}")

        try:
            response = self.session.post(
                f"{self.base_url}/rest",
                headers=self.headers,
                json=payload,
                timeout=self.timeout,
                verify=True,
                allow_redirects=False,
            )
        except requests.RequestException as error:
            raise BridgeError(f"ThirdEye API call {method} failed to connect.") from error

        if response.is_redirect:
            raise BridgeError(f"ThirdEye API call {method} was redirected.")
        try:
            response.raise_for_status()
        except requests.HTTPError as error:
            raise BridgeError(
                f"ThirdEye API call {method} failed with HTTP {response.status_code}."
            ) from error

        try:
            data = response.json()
        except requests.exceptions.JSONDecodeError as error:
            raise BridgeError(f"ThirdEye API call {method} returned invalid JSON.") from error
        if data.get("error"):
            raise BridgeError(f"ThirdEye API call {method} failed: {data['error']}")
        return data.get("result")

    def inventory_addresses(self, network, page_size=500):
        addresses = set()
        offset = 0

        while True:
            page = self.call(
                "Inventory.search",
                network=[network],
                scheme="ipAddress",
                query="\n",
                pageData={"offset": offset, "pageSize": page_size},
                sortColumn="ipAddress",
                descending=False,
            ) or {}
            devices = page.get("devices") or []
            for device in devices:
                candidate = device.get("ipAddress")
                try:
                    addresses.add(str(ipaddress.ip_address(candidate)))
                except (ValueError, TypeError):
                    continue

            offset += len(devices)
            total = page.get("total", offset)
            if not devices or offset >= total:
                return addresses

    def find_job(self, network, job_name, page_size=100):
        offset = 0
        matches = []

        while True:
            page = self.call(
                "Scheduler.searchJobs",
                pageData={
                    "offset": offset,
                    "jobData": [],
                    "pageSize": page_size,
                    "total": 1,
                },
                networks=[network],
                sortColumn="",
                descending=False,
            ) or {}
            jobs = page.get("jobData") or []
            matches.extend(job for job in jobs if job.get("jobName") == job_name)
            offset += len(jobs)
            total = page.get("total", offset)
            if not jobs or offset >= total:
                break

        if not matches:
            raise BridgeError(f'No available job named "{job_name}" was found.')
        if len(matches) > 1:
            ids = ", ".join(str(job.get("jobId")) for job in matches)
            raise BridgeError(f'Multiple jobs named "{job_name}" were found: {ids}')

        job_id = matches[0].get("jobId")
        job_data = self.call("Scheduler.getJob", jobId=job_id)
        if job_data is None:
            raise BridgeError(f"Scheduler.getJob returned no data for job ID {job_id}.")
        return job_data

    def run_now(self, job_data):
        return self.call("Scheduler.runNow", jobData=job_data)


def parse_args(argv):
    parser = argparse.ArgumentParser(
        description="Find LiveNX devices missing from ThirdEye inventory."
    )
    parser.add_argument(
        "--apply",
        action="store_true",
        help="Run the configured ThirdEye discovery job for the missing addresses.",
    )
    return parser.parse_args(argv)


def main(argv=None):
    args = parse_args(argv)
    config = BridgeConfig.from_env()

    live_nx = LiveNXClient(
        config.live_nx_base_url,
        config.live_nx_api_token,
        config.live_nx_export_path,
        config.timeout,
    )
    netld = NetLDClient(
        config.netld_base_url,
        config.netld_api_key,
        config.timeout,
        config.debug,
    )

    live_nx_addresses = parse_live_nx_device_ips(
        live_nx.export_devices_csv(),
        require_vendor=config.require_vendor,
    )
    netld.login()
    managed_addresses = netld.inventory_addresses(config.netld_network)
    missing_addresses = live_nx_addresses - managed_addresses

    print(f"LiveNX device addresses: {len(live_nx_addresses)}")
    print(f"ThirdEye managed addresses: {len(managed_addresses)}")
    print(f"Missing from ThirdEye: {len(missing_addresses)}")
    for address in sorted_ip_addresses(missing_addresses):
        print(f"  {address}")

    if not missing_addresses:
        print("No discovery is required.")
        return 0
    if not args.apply:
        print("Dry run only. Re-run with --apply to start discovery.")
        return 0
    if not config.discovery_job_name:
        raise BridgeError(
            "Set NETLD_DISCOVERY_JOB_NAME to an existing Discover Devices job "
            "before using --apply."
        )

    job_data = netld.find_job(
        config.netld_network,
        config.discovery_job_name,
    )
    prepared_job = prepare_discovery_job(
        job_data,
        config.netld_network,
        missing_addresses,
    )
    execution = netld.run_now(prepared_job)
    print("Discovery started:")
    print(json.dumps(execution, indent=2))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except BridgeError as error:
        print(f"Error: {error}", file=sys.stderr)
        raise SystemExit(1)

File: Integrations/LiveAction/LiveNX-Inventory-To-ThirdEye/nodeJS/live-nx-to-thirdeye.mjs
#!/usr/bin/env node

import "dotenv/config";

import { randomUUID } from "node:crypto";
import { isIP } from "node:net";
import process from "node:process";
import { pathToFileURL } from "node:url";

import { parse } from "csv-parse/sync";


export class BridgeError extends Error {}


export function environmentBoolean(name, defaultValue = false) {
  const value = process.env[name];
  if (value === undefined || value.trim() === "") {
    return defaultValue;
  }
  return ["1", "true", "yes", "on"].includes(value.trim().toLowerCase());
}


function requiredEnvironmentValue(name) {
  const value = process.env[name];
  if (!value) {
    throw new BridgeError(`Set ${name} in .env before running this integration.`);
  }
  return value;
}


function normalizedHeader(value) {
  return value.toUpperCase().replaceAll(/[^A-Z0-9]/g, "");
}


export function canonicalIpAddress(value) {
  const candidate = String(value || "").trim();
  const version = isIP(candidate);
  if (version === 4) {
    return candidate
      .split(".")
      .map((part) => Number.parseInt(part, 10))
      .join(".");
  }
  if (version === 6) {
    return new URL(`http://[${candidate}]/`).hostname.slice(1, -1);
  }
  return null;
}


export function sortedIpAddresses(addresses) {
  return [...addresses].sort((left, right) => {
    const versionDifference = isIP(left) - isIP(right);
    return versionDifference || left.localeCompare(right, "en", { numeric: true });
  });
}


export function parseLiveNXDeviceIps(csvText, { requireVendor = true } = {}) {
  let rows;
  try {
    rows = parse(csvText, {
      bom: true,
      columns: true,
      relax_column_count: true,
      skip_empty_lines: true,
      trim: true,
    });
  } catch (error) {
    throw new BridgeError(`Could not parse the LiveNX device CSV: ${error.message}`);
  }
  if (rows.length === 0) {
    return new Set();
  }

  const fields = new Map(
    Object.keys(rows[0]).map((name) => [normalizedHeader(name), name]),
  );
  const ipField = ["IPADDRESS", "MANAGEMENTIPADDRESS", "MANAGEMENTIP", "IP"]
    .map((name) => fields.get(name))
    .find(Boolean);
  if (!ipField) {
    throw new BridgeError(
      `The LiveNX CSV does not contain a recognized IP-address column. Available columns: ${Object.keys(
        rows[0],
      ).join(", ")}`,
    );
  }

  const vendorField = fields.get("VENDOR");
  if (requireVendor && !vendorField) {
    throw new BridgeError(
      "LIVENX_REQUIRE_VENDOR=true, but the LiveNX CSV has no VENDOR column.",
    );
  }

  const addresses = new Set();
  for (const row of rows) {
    if (requireVendor && !String(row[vendorField] || "").trim()) {
      continue;
    }
    const address = canonicalIpAddress(row[ipField]);
    if (address) {
      addresses.add(address);
    }
  }
  return addresses;
}


export function prepareDiscoveryJob(jobData, network, addresses) {
  const prepared = structuredClone(jobData);
  const parameters = prepared?.jobParameters;
  if (!parameters || typeof parameters !== "object" || Array.isArray(parameters)) {
    throw new BridgeError("The selected job has no jobParameters object.");
  }
  if (!("includedAddresses" in parameters)) {
    throw new BridgeError(
      "The selected job is not a compatible Discover Devices job: " +
        "jobParameters.includedAddresses is missing.",
    );
  }

  if ("managedNetwork" in prepared) {
    prepared.managedNetwork = network;
  }
  if ("managedNetworks" in prepared) {
    prepared.managedNetworks = Array.isArray(prepared.managedNetworks) ? [network] : network;
  }
  if ("managedNetwork" in parameters) {
    parameters.managedNetwork = network;
  }
  parameters.includedAddresses = sortedIpAddresses(addresses).join(",");
  return prepared;
}


function isRedirect(response) {
  return response.status >= 300 && response.status < 400;
}


function splitSetCookie(header) {
  return header ? header.split(/,(?=\s*[^;=]+=[^;]+)/) : [];
}


export class LiveNXClient {
  constructor(baseUrl, apiToken, exportPath, timeout = 30000, fetchFunction = fetch) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiToken = apiToken;
    this.exportPath = `/${exportPath.replace(/^\//, "")}`;
    this.timeout = timeout;
    this.fetch = fetchFunction;
  }

  async exportDevicesCsv() {
    const url = `${this.baseUrl}${this.exportPath}`;
    let response;
    try {
      response = await this.fetch(url, {
        headers: {
          Accept: "text/csv",
          Authorization: `Bearer ${this.apiToken}`,
        },
        redirect: "manual",
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch (error) {
      throw new BridgeError(`Could not retrieve the LiveNX device export from ${url}.`);
    }
    if (isRedirect(response)) {
      throw new BridgeError(
        "The LiveNX export request was redirected. Confirm LIVENX_BASE_URL, " +
          "LIVENX_DEVICE_EXPORT_PATH, and the API token.",
      );
    }
    if (!response.ok) {
      throw new BridgeError(`LiveNX device export failed with HTTP ${response.status}.`);
    }
    return response.text();
  }
}


export class NetLDClient {
  constructor(baseUrl, apiKey, timeout = 30000, debug = false, fetchFunction = fetch) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.timeout = timeout;
    this.debug = debug;
    this.fetch = fetchFunction;
    this.cookieHeader = "";
  }

  headers(extra = {}) {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
      ...(this.cookieHeader ? { Cookie: this.cookieHeader } : {}),
      ...extra,
    };
  }

  updateCookies(response) {
    const setCookie = response.headers.getSetCookie
      ? response.headers.getSetCookie()
      : splitSetCookie(response.headers.get("set-cookie") || "");
    this.cookieHeader = setCookie
      .map((cookie) => cookie.split(";")[0])
      .filter(Boolean)
      .join("; ");
  }

  async login() {
    let response;
    try {
      response = await this.fetch(`${this.baseUrl}/rest`, {
        headers: { Authorization: `Bearer ${this.apiKey}` },
        redirect: "manual",
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch (error) {
      throw new BridgeError(`Could not reach ${this.baseUrl}.`);
    }
    if (isRedirect(response)) {
      throw new BridgeError("ThirdEye login redirected instead of creating an API session.");
    }
    if (!response.ok) {
      throw new BridgeError(`ThirdEye login failed with HTTP ${response.status}.`);
    }
    this.updateCookies(response);
  }

  async call(method, params = {}) {
    const payload = {
      jsonrpc: "2.0",
      method,
      params,
      id: randomUUID(),
    };
    if (this.debug) {
      console.log(`ThirdEye request: ${JSON.stringify(payload, null, 2)}`);
    }

    let response;
    try {
      response = await this.fetch(`${this.baseUrl}/rest`, {
        method: "POST",
        headers: this.headers(),
        body: JSON.stringify(payload),
        redirect: "manual",
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch (error) {
      throw new BridgeError(`ThirdEye API call ${method} failed to connect.`);
    }
    if (isRedirect(response)) {
      throw new BridgeError(`ThirdEye API call ${method} was redirected.`);
    }
    if (!response.ok) {
      throw new BridgeError(
        `ThirdEye API call ${method} failed with HTTP ${response.status}.`,
      );
    }

    let data;
    try {
      data = await response.json();
    } catch (error) {
      throw new BridgeError(`ThirdEye API call ${method} returned invalid JSON.`);
    }
    if (data.error) {
      throw new BridgeError(`ThirdEye API call ${method} failed: ${JSON.stringify(data.error)}`);
    }
    return data.result;
  }

  async inventoryAddresses(network, pageSize = 500) {
    const addresses = new Set();
    let offset = 0;
    while (true) {
      const page =
        (await this.call("Inventory.search", {
          network: [network],
          scheme: "ipAddress",
          query: "\n",
          pageData: { offset, pageSize },
          sortColumn: "ipAddress",
          descending: false,
        })) || {};
      const devices = page.devices || [];
      for (const device of devices) {
        const address = canonicalIpAddress(device.ipAddress);
        if (address) {
          addresses.add(address);
        }
      }
      offset += devices.length;
      const total = page.total ?? offset;
      if (devices.length === 0 || offset >= total) {
        return addresses;
      }
    }
  }

  async findJob(network, jobName, pageSize = 100) {
    const matches = [];
    let offset = 0;
    while (true) {
      const page =
        (await this.call("Scheduler.searchJobs", {
          pageData: { offset, jobData: [], pageSize, total: 1 },
          networks: [network],
          sortColumn: "",
          descending: false,
        })) || {};
      const jobs = page.jobData || [];
      matches.push(...jobs.filter((job) => job.jobName === jobName));
      offset += jobs.length;
      const total = page.total ?? offset;
      if (jobs.length === 0 || offset >= total) {
        break;
      }
    }
    if (matches.length === 0) {
      throw new BridgeError(`No available job named "${jobName}" was found.`);
    }
    if (matches.length > 1) {
      throw new BridgeError(
        `Multiple jobs named "${jobName}" were found: ${matches
          .map((job) => job.jobId)
          .join(", ")}`,
      );
    }

    const job = await this.call("Scheduler.getJob", { jobId: matches[0].jobId });
    if (!job) {
      throw new BridgeError(
        `Scheduler.getJob returned no data for job ID ${matches[0].jobId}.`,
      );
    }
    return job;
  }

  runNow(jobData) {
    return this.call("Scheduler.runNow", { jobData });
  }
}


function configurationFromEnvironment() {
  const timeoutSeconds = Number.parseFloat(process.env.REQUEST_TIMEOUT_SECONDS || "30");
  if (!Number.isFinite(timeoutSeconds) || timeoutSeconds <= 0) {
    throw new BridgeError("REQUEST_TIMEOUT_SECONDS must be a number greater than zero.");
  }
  return {
    liveNXBaseUrl: requiredEnvironmentValue("LIVENX_BASE_URL"),
    liveNXApiToken: requiredEnvironmentValue("LIVENX_API_TOKEN"),
    liveNXExportPath: process.env.LIVENX_DEVICE_EXPORT_PATH || "/v1/devices/export/csv",
    requireVendor: environmentBoolean("LIVENX_REQUIRE_VENDOR", true),
    netLDBaseUrl: requiredEnvironmentValue("NETLD_BASE_URL"),
    netLDApiKey: requiredEnvironmentValue("NETLD_API_KEY"),
    network: process.env.NETLD_NETWORK || "Default",
    discoveryJobName: process.env.NETLD_DISCOVERY_JOB_NAME,
    debug: environmentBoolean("NETLD_DEBUG"),
    timeout: timeoutSeconds * 1000,
  };
}


export async function runBridge({ apply = false } = {}) {
  const config = configurationFromEnvironment();
  const liveNX = new LiveNXClient(
    config.liveNXBaseUrl,
    config.liveNXApiToken,
    config.liveNXExportPath,
    config.timeout,
  );
  const netLD = new NetLDClient(
    config.netLDBaseUrl,
    config.netLDApiKey,
    config.timeout,
    config.debug,
  );

  const liveNXAddresses = parseLiveNXDeviceIps(await liveNX.exportDevicesCsv(), {
    requireVendor: config.requireVendor,
  });
  await netLD.login();
  const managedAddresses = await netLD.inventoryAddresses(config.network);
  const missingAddresses = new Set(
    [...liveNXAddresses].filter((address) => !managedAddresses.has(address)),
  );

  console.log(`LiveNX device addresses: ${liveNXAddresses.size}`);
  console.log(`ThirdEye managed addresses: ${managedAddresses.size}`);
  console.log(`Missing from ThirdEye: ${missingAddresses.size}`);
  for (const address of sortedIpAddresses(missingAddresses)) {
    console.log(`  ${address}`);
  }

  if (missingAddresses.size === 0) {
    console.log("No discovery is required.");
    return;
  }
  if (!apply) {
    console.log("Dry run only. Re-run with --apply to start discovery.");
    return;
  }
  if (!config.discoveryJobName) {
    throw new BridgeError(
      "Set NETLD_DISCOVERY_JOB_NAME to an existing Discover Devices job before using --apply.",
    );
  }

  const job = await netLD.findJob(config.network, config.discoveryJobName);
  const prepared = prepareDiscoveryJob(job, config.network, missingAddresses);
  const execution = await netLD.runNow(prepared);
  console.log("Discovery started:");
  console.log(JSON.stringify(execution, null, 2));
}


function showHelp() {
  console.log("Usage: node live-nx-to-thirdeye.mjs [--apply]");
  console.log("");
  console.log("Find LiveNX devices missing from ThirdEye inventory.");
  console.log("  --apply  Run the configured discovery job for missing addresses.");
}


async function main() {
  const unknown = process.argv.slice(2).filter((value) => !["--apply", "--help", "-h"].includes(value));
  if (unknown.length > 0) {
    throw new BridgeError(`Unknown argument: ${unknown[0]}`);
  }
  if (process.argv.includes("--help") || process.argv.includes("-h")) {
    showHelp();
    return;
  }
  await runBridge({ apply: process.argv.includes("--apply") });
}


if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((error) => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
}
